src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -1090,10 +1108,42 @@ - (void)_reportAttributes:(NSArray<NSDictionary<NSString *, id> *> *)attributes
     }
 }
 
+// When we receive an attribute report, check if there are any changes in parts list, server list, device type list, accepted commands list,
+//  attribute list, cluster revision or feature map attributes of the descriptor cluster. If yes, make a note that the device configuration changed.
+- (void)_noteDeviceConfigurationChanged:(NSArray<NSDictionary<NSString *, id> *> *)attributeReport
+{
+    for (NSDictionary<NSString *, id> * attribute in attributeReport) {
+        MTRAttributePath * attributePath = attribute[MTRAttributePathKey];
+
+        if (attributePath.cluster.unsignedLongValue != MTRClusterDescriptorID) {
+            return;
+        }
+
+        switch (attributePath.attribute.unsignedLongValue) {
+        case MTRClusterDescriptorAttributePartsListID:
+        case MTRClusterDescriptorAttributeServerListID:
+        case MTRClusterDescriptorAttributeDeviceTypeListID:
+        case MTRClusterDescriptorAttributeAcceptedCommandListID:
+        case MTRClusterDescriptorAttributeAttributeListID:
+        case MTRClusterDescriptorAttributeClusterRevisionID:
+        case MTRClusterDescriptorAttributeFeatureMapID: {
+            // If changes are detected, note that the device configuration has changed.
+            NSDictionary * cachedAttributeDataValue = [self _cachedAttributeValueForPath:attributePath];
+            if (cachedAttributeDataValue != nil && ![self _attributeDataValue:attribute[MTRDataKey] isEqualToDataValue:cachedAttributeDataValue]) {
+                _deviceConfigurationChanged = YES;
+                break;
+            }
+        }
+        }
+    }
+}
+
 - (void)_handleAttributeReport:(NSArray<NSDictionary<NSString *, id> *> *)attributeReport
 {
     std::lock_guard lock(_lock);
 
+    [self _noteDeviceConfigurationChanged:attributeReport];
```

Comment:
One more thing worth thinking about: in `_getAttributesToReportWithReportedValues` we are already doing the "did the value change from our cache" comparisons, right?   Perhaps in that code, if the value has changed, we should check whether the path is one of the ones we care about and set the flag?

That would incidentally also fix the fact that the "value has changed" check this PR implements is wrong, but the way the test is written, by injecting a value that does not match the values we would actually get passed to `_handleAttributeReport` does not catch that.

Which leads me to a request: can you please write a test that would actually test this properly?  Followup is OK for that, but the test here is clearly not it.

Reply:
And also: the test does not have tests that the "configuration changed" callback _is not called_ when the configuration has in fact not changed.  Those tests need to exist too, and would have caught the "value has changed" check being wrong.

Reply:
This makes sense. added a test to the test017 to check if the device configuration changed callback is not called